### DISCUSSION TITLE
QDate - How to use events_date with js function?

### DISCUSSION THREAD

**Discussion by docm64** [elimintz/justpy#250] [Q&A]
Quasar's QDate class has a property events (in JP it's called events_date to avoid name conflict with built-in events property)
According to Quasar documentation, events can be either an array of date strings or a js function.
This one works in JustPy as expected:
```python
    dt = jp.QDate()
    dt.events_date = str(['2021/03/30', '2021/03/31'])
```
However, this doesn't work (should mark all even days, but doesn't mark any)
```python
    dt = jp.QDate()
    dt.events_date = "date => date[9] % 2 === 0"
```
I cannot figure out my mistake and didn't find anything in the docs about how to use function-style properties of Quasar classes.
Any hint?

**Comment by elimintz** [comment]
Function style properties of Quasar classes are not supported.
This is not what you want exactly but maybe could work:
```python
import justpy as jp
import datetime


def set_events(date):
    date_list = []
    for i in range(-30, 30):
        date_i = date + datetime.timedelta(days=i)
        if date_i.day % 2 == 0:
            date_list.append(date_i.strftime('%Y/%m/%d'))
    return str(date_list)


def date_test():
    wp = jp.QuasarPage()
    dt = jp.QDate(a=wp, style='margin: 40px')
    dt.events_date = set_events(datetime.datetime.now())
    return wp


jp.justpy(date_test)
```

**Reply by docm64** [reply]
Hi Elimintz,
I had created a workaround as you described and it works in the case of events_data. But unfortunately it doesn't solve my real issue. I want to assign different color codes to specific events using the event_color property of QDate which requires a function-style property. No array support is available for event_color.
Sorry to hear that function style properties are not supported. Was wondering why. Is there a way to workaround this limitation?

**Comment by elimintz** [comment]
In principle, it should work as an evaluated_prop. That is how I implemented rules in QInput (https://quasar.dev/vue-components/input#internal-validation). If you add a prop to the evaluated_prop list it gets evaluated (see the python implementation of QInput  and the quasar_component.js ). For some reason this does not work with events_date. I would need to look into the Quasar implementation to figure out why.

**Comment by docm64** [comment]
I just looked at the code in quasar_components.js and I think there is a bug in line 28. You use j as index variable for the inner loop but then compare/increment i which is the index variable of the outer loop. Please doublecheck.

**Reply by elimintz** [reply]
Yes, thank you for catching that!

**Comment by docm64** [comment]
In order to use a function-style property with QDate, the following is necessary: 
```python
    qd = jp.QDate()
    qd.events_date = "(date) => {return date[9] %3 === 0}"
    qd.evaluate_prop.append('events')
```
This made it work for me.

Note that the mapping of property names is a little bit tricky. In Quasar, the property is named QDate.events.
Due to the fact that 'events' is a reserved property in JustPy, the property is called events_date in the Python domain.

However the evaluate_prop list is only analyzed in the JS domain, where the property name is 'events'

**Reply by elimintz** [reply]
Thanks, very helpful. I forgot that I had this problem with QDate when I wrote the code...
